src/Core/Domain/Product/Image/QueryResult/Shop/ShopProductImages.php

```diff
@@ -0,0 +1,61 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Domain\Product\Image\QueryResult\Shop;
+
+class ShopProductImages
```

Comment:
I wonder if maybe `ShopProductImages` could be the collection itself? And of course adding it the additional `shopId` field This would save us one class (`ProductImageCollection` wouldn't be needed)

Reply:
```suggestion
/**
 * @template-extends ImmutableCollection<ShopProductImages>
 */
class ShopProductImages extends ImmutableCollection
```

Reply:
We can't since there is special behaviours on the ImmutableCollection which imply to have the same behaviour when we build collection. For example, have a look on the filter function in ImmutableCollection, this function return a new static(array) not static(array, int) or something else